swaybar/tray/dbusmenu.c

```diff
@@ -0,0 +1,1381 @@
+		struct swaybar_dbusmenu_menu *child_menu = menu->child_menus->items[j];
+		if (child_menu->item_id == item_id) {
+			close_menus(child_menu);
+		}
+	}
+}
+
+static void close_child_menus_outside_pointer(struct swaybar_dbusmenu_menu *menu,
+											  struct swaybar_seat *seat) {
+	for (int i = 0; i < menu->items->length; ++i) {
+		struct swaybar_dbusmenu_menu_item *item = menu->items->items[i];
+
+		int scale = menu->dbusmenu->output->scale;
+		int x = seat->pointer.x * scale;
+		int y = seat->pointer.y * scale;
+		if (item->submenu && !is_in_hotspot(&item->hotspot, x, y)) {
+		  close_menus_id(menu, item->id);
+		}
+	}
+}
+
+bool dbusmenu_pointer_frame(struct swaybar_seat *data, 
+							struct wl_pointer *wl_pointer) {
+	struct swaybar_tray *tray = data->bar->tray;
+	if (!(tray && tray->menu && tray->menu_pointer_focus)) {
+		return false;
+	}
+	return true;
+}
+
+bool dbusmenu_pointer_axis(struct swaybar_seat *data,
+                           struct wl_pointer *wl_pointer) {
+	struct swaybar_tray *tray = data->bar->tray;
+	if (!(tray && tray->menu && tray->menu_pointer_focus)) {
+		return false;
+	}
+	return true;
+}
+
+bool dbusmenu_pointer_enter(void *data, struct wl_pointer *wl_pointer,
+                            uint32_t serial, struct wl_surface *surface,
+                            wl_fixed_t surface_x, wl_fixed_t surface_y) {
+	struct swaybar_seat *seat = data;
+	struct swaybar_tray *tray = seat->bar->tray;
+	if (!(tray && tray->menu)) {
+		return false;
+	}
+
+	struct swaybar_dbusmenu_menu *new_focused_menu =
+	dbusmenu_menu_find_menu_surface(tray->menu->menu, surface);
+
+	if (new_focused_menu && new_focused_menu->child_menus) {
+		close_child_menus_outside_pointer(new_focused_menu, seat);
+	}
+
+	tray->menu_pointer_focus = new_focused_menu;
+
+	return true;
+}
+
+bool dbusmenu_pointer_leave(void *data, struct wl_pointer *wl_pointer,
+                            uint32_t serial, struct wl_surface *surface) {
+	struct swaybar_seat *seat = data;
+	struct swaybar_tray *tray = seat->bar->tray;
+	if (!(tray && tray->menu)) {
+		return false;
+	}
+
+	tray->menu_pointer_focus = NULL;
+
+	return true;
+}
+
+static bool dbusmenu_pointer_button_left_process_item(struct swaybar_dbusmenu *dbusmenu,
+													  struct swaybar_dbusmenu_menu_item *item,
+													  struct swaybar_seat *seat) {
+	struct swaybar_sni *sni = dbusmenu->sni;
+	struct swaybar_tray *tray = sni->tray;
+	int scale = dbusmenu->output->scale;
+
+	if (is_in_hotspot(&item->hotspot, seat->pointer.x * scale,
+			seat->pointer.y * scale)) {
+		if (!item->enabled || item->is_separator) {
+		  return false;
+		}
+
+		sway_log(SWAY_DEBUG, "%s%s menu clicked id %d", sni->service, sni->menu,
+			item->id);
+
+		sd_bus_call_method_async(tray->bus, NULL, sni->service, sni->menu,
+			menu_interface, "Event", NULL, NULL, "isvu", item->id, "clicked", "y", 0, 
+			time(NULL));
+
+		if (!tray->menu->drawing) {
+			swaybar_dbusmenu_destroy(tray->menu);
```

Comment:
Why do we destroy the menu if we are not drawing it?

Reply:
The menu was only allowed to be destroyed if no drawing was ongoing in parallel, I removed the drawing variable as it now not necessary anymore.